Sources/SwiftDocCUtilities/Action/Actions/Convert/JSONEncodingRenderNodeWriter.swift

```diff
@@ -73,49 +73,88 @@ class JSONEncodingRenderNodeWriter {
         
         let renderNodeTargetFolderURL = renderNodeTargetFileURL.deletingLastPathComponent()
         
-        // On Linux sometimes it takes a moment for the directory to be created and that leads to
-        // errors when trying to write files concurrently in the same target location.
-        // We keep an index in `directoryIndex` and create new sub-directories as needed.
-        // When the symbol's directory already exists no code is executed during the lock below
-        // besides the set lookup.
-        try directoryIndex.sync { directoryIndex in
-            let (insertedRenderNodeTargetFolderURL, _) = directoryIndex.insert(renderNodeTargetFolderURL)
-            if insertedRenderNodeTargetFolderURL {
+        let queue = DispatchQueue(
+            label: "org.swift.JSONEncodingRenderNodeWriterQueue",
+            qos: .unspecified,
+            attributes: .concurrent
+        )
+        let group = DispatchGroup()
+        
+        var renderJSONWritingError: Swift.Error? = nil
+        
+        group.enter()
+        queue.async { [self] in
```

Comment:
I had assumed this would be slower because of the synchronization overhead as well but I actually saw a significant performance gain in practice. Here's the benchmark results for just this change:

```
TestFramework-5
+-----------------------------------------------------------------------------------------------------+
| Metric                                   | Change     | Before               | After                |
+-----------------------------------------------------------------------------------------------------+
| Duration for 'bundle-registration' (sec) | -0.91%     | 2.65                 | 2.62                 |
| Duration for 'convert-total-time' (sec)  | -1.62%     | 4.20                 | 4.13                 |
| Duration for 'documentation-processing'  | -3.34%     | 1.38                 | 1.33                 |
| Duration for 'finalize-navigation-index' | no change  | 0.08                 | 0.08                 |
| Peak memory footprint (MB)               | +9.38%     | 187.55               | 205.15               |
| Data subdirectory size (MB)              | no change  | 57.36                | 57.36                |
| Index subdirectory size (MB)             | no change  | 0.47                 | 0.47                 |
| Total DocC archive size (MB)             | no change  | 74.45                | 74.45                |
| Topic Anchor Checksum                    | no change  | 65d4ff3050cd1106a21b | 65d4ff3050cd1106a21b |
| Topic Graph Checksum                     | no change  | 0a64c740a2ed5a246836 | 0a64c740a2ed5a246836 |
+-----------------------------------------------------------------------------------------------------+
TestFramework-10
+-----------------------------------------------------------------------------------------------------+
| Metric                                   | Change     | Before               | After                |
+-----------------------------------------------------------------------------------------------------+
| Duration for 'bundle-registration' (sec) | +0.23%     | 5.24                 | 5.26                 |
| Duration for 'convert-total-time' (sec)  | -2.42%     | 8.59                 | 8.38                 |
| Duration for 'documentation-processing'  | -7.32%     | 3.02                 | 2.80                 |
| Duration for 'finalize-navigation-index' | +1.35%     | 0.15                 | 0.15                 |
| Peak memory footprint (MB)               | +6.51%     | 341.62               | 363.85               |
| Data subdirectory size (MB)              | no change  | 114.67               | 114.67               |
| Index subdirectory size (MB)             | no change  | 0.95                 | 0.95                 |
| Total DocC archive size (MB)             | no change  | 148.04               | 148.04               |
| Topic Anchor Checksum                    | no change  | c8f03d4e81da8e3b6d7f | c8f03d4e81da8e3b6d7f |
| Topic Graph Checksum                     | no change  | edce22bce4f8de475a5e | edce22bce4f8de475a5e |
+-----------------------------------------------------------------------------------------------------+
TestFramework-25
+-----------------------------------------------------------------------------------------------------+
| Metric                                   | Change     | Before               | After                |
+-----------------------------------------------------------------------------------------------------+
| Duration for 'bundle-registration' (sec) | +0.38%     | 13.37                | 13.42                |
| Duration for 'convert-total-time' (sec)  | -0.92%     | 22.29                | 22.08                |
| Duration for 'documentation-processing'  | -3.21%     | 8.12                 | 7.86                 |
| Duration for 'finalize-navigation-index' | +2.19%     | 0.36                 | 0.37                 |
| Peak memory footprint (MB)               | +1.91%     | 826.78               | 842.57               |
| Data subdirectory size (MB)              | no change  | 288.21               | 288.21               |
| Index subdirectory size (MB)             | no change  | 2.41                 | 2.41                 |
| Total DocC archive size (MB)             | no change  | 370.43               | 370.43               |
| Topic Anchor Checksum                    | no change  | 9a675b9ad6d69f8b7f0c | 9a675b9ad6d69f8b7f0c |
| Topic Graph Checksum                     | no change  | 665713509084b5131a37 | 665713509084b5131a37 |
+-----------------------------------------------------------------------------------------------------+
TestFramework-50
+-----------------------------------------------------------------------------------------------------+
| Metric                                   | Change     | Before               | After                |
+-----------------------------------------------------------------------------------------------------+
| Duration for 'bundle-registration' (sec) | no change  | 26.96                | 26.96                |
| Duration for 'convert-total-time' (sec)  | -3.42%     | 48.12                | 46.48                |
| Duration for 'documentation-processing'  | -8.20%     | 19.46                | 17.86                |
| Duration for 'finalize-navigation-index' | -1.19%     | 0.75                 | 0.74                 |
| Peak memory footprint (MB)               | +0.86%     | 1586.90              | 1600.50              |
| Data subdirectory size (MB)              | no change  | 578.76               | 578.76               |
| Index subdirectory size (MB)             | no change  | 4.84                 | 4.84                 |
| Total DocC archive size (MB)             | no change  | 742.42               | 742.42               |
| Topic Anchor Checksum                    | no change  | 08d0a84bec913460905f | 08d0a84bec913460905f |
| Topic Graph Checksum                     | no change  | 74d25c4f1ee185ad5676 | 74d25c4f1ee185ad5676 |
+-----------------------------------------------------------------------------------------------------+
```

Reply:
That is the opposite of what I'm seeing:
```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Metric                                   │ Change          │ 200b9d9fa7           │ 17334f32a7           │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Duration for 'bundle-registration'       │ no change¹,²    │ 3.243 sec            │ 3.217 sec            │
│ Duration for 'convert-total-time'        │ +3.058%³,⁴      │ 5.341 sec            │ 5.504 sec            │
│ Duration for 'documentation-processing'  │ +9.535%⁵,⁶      │ 1.99 sec             │ 2.179 sec            │
│ Duration for 'finalize-navigation-index' │ no change⁷      │ 0.049 sec            │ 0.049 sec            │
│ Peak memory footprint                    │ no change⁸      │ 412.4 MB             │ 410.9 MB             │
│ Data subdirectory size                   │ no change       │ 111.9 MB             │ 111.9 MB             │
│ Index subdirectory size                  │ no change       │ 994 KB               │ 994 KB               │
│ Total DocC archive size                  │ no change       │ 149.4 MB             │ 149.4 MB             │
│ Topic Anchor Checksum                    │ no change       │ c8f03d4e81da8e3b6d7f │ c8f03d4e81da8e3b6d7f │
│ Topic Graph Checksum                     │ no change       │ edce22bce4f8de475a5e │ edce22bce4f8de475a5e │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
...
 3: There's a statistically significant difference between the two benchmark measurements.
    The values are different enough that the most probable explanation is that they're random samples from two different data sets.
    t-statistic : -2.915          degrees of freedom : 18       95% confidence critical value : +2.086

 4: A human should check that the measurements for this metric look like random samples around a certain value to ensure the validity of this result.
...
```
The `17334f32a7` values in my measurement were marked as potentially biased but look reasonably normal distributed to me:
```
The 'duration-convert-total-time' samples from the 'after' measurements show possible signs of bias.
The benchmark diff analysis assumes that these values follow a normal distribution around the value that they're meant to represent. If they don't the conclusion from the diff analysis could be invalid.
A human should inspect these values. If the samples look biased or look too noisy you can try gathering new metrics with more 'repetitions' or with a larger .docc catalog.
Full precision values: [5.70400, 5.76700, 5.46400, 5.47900, 5.50700, 5.47300, 5.45600, 5.39700, 5.46900, 5.32600]
████████████████████████████████████████████████████████████████████████████████   5.704 sec
█████████████████████████████████████████████████████████████████████████████████   5.767 sec
████████████████████████████████████████████████████████████████████████████▊   5.464 sec
████████████████████████████████████████████████████████████████████████████▉   5.479 sec
█████████████████████████████████████████████████████████████████████████████▎   5.507 sec
████████████████████████████████████████████████████████████████████████████▉   5.473 sec
████████████████████████████████████████████████████████████████████████████▋   5.456 sec
███████████████████████████████████████████████████████████████████████████▊   5.397 sec
████████████████████████████████████████████████████████████████████████████▊   5.469 sec
██████████████████████████████████████████████████████████████████████████▊   5.326 sec
```
Even without the first two metrics the average is higher.

Reply:
Can you checkout https://github.com/apple/swift-docc/pull/159 and run these two commands to see a more detailed analysis of the differences?
```
swift run benchmark measure-commits <this commit hash> <the commit before> --repetitions 15 --docc-arguments convert path/to/test-bundle-of-at-least-size10.docc
```
and then 
```
swift run benchmark diff benchmark-<the commit beofre>.json benchmark-<this commit>.json
```

Reply:
Oh interesting... Maybe this is CPU specific? These benchmarks were on a 6-core (3.5 ghz) machine: https://support.apple.com/kb/SP697.

Reply:
I'm seeing a performance improvement there too... This is now running on an 8-core iMac Pro.

This was my invocation:

```shell
swift run --package-path bin/benchmark benchmark measure-commits 82b1d4aeb2fb4afd8d5bc09893ac07b517999d89 ed63fcff152f11278632434b7f5d0636803b4403  --repetitions 15 --docc-arguments convert /path/to/Docs.docc

swift run --package-path bin/benchmark benchmark diff benchmark-82b1d4aeb2fb4afd8d5bc09893ac07b517999d89.json benchmark-ed63fcff152f11278632434b7f5d0636803b4403.json
```

And the results:

```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Metric                                   │ Change          │ 82b1d4aeb2fb4afd8d5bc09893ac07b517999d89 │ ed63fcff152f11278632434b7f5d0636803b4403 │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Duration for 'bundle-registration'       │ -5.231%¹        │ 4.488 sec                                │ 4.254 sec                                │
│ Duration for 'convert-total-time'        │ -5.377%²        │ 7.443 sec                                │ 7.042 sec                                │
│ Duration for 'documentation-processing'  │ -5.21%³         │ 2.71 sec                                 │ 2.569 sec                                │
│ Duration for 'finalize-navigation-index' │ -5.209%⁴        │ 0.11 sec                                 │ 0.104 sec                                │
│ Peak memory footprint                    │ -1.103%⁵,⁶      │ 436.6 MB                                 │ 431.7 MB                                 │
│ Data subdirectory size                   │ no change       │ 111.9 MB                                 │ 111.9 MB                                 │
│ Index subdirectory size                  │ no change       │ 994 KB                                   │ 994 KB                                   │
│ Total DocC archive size                  │ no change       │ 149.4 MB                                 │ 149.4 MB                                 │
│ Topic Anchor Checksum                    │ no change       │ c8f03d4e81da8e3b6d7fb7f63c771147         │ c8f03d4e81da8e3b6d7fb7f63c771147         │
│ Topic Graph Checksum                     │ no change       │ edce22bce4f8de475a5e58a2e45dcdc0         │ edce22bce4f8de475a5e58a2e45dcdc0         │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

 1: There's a statistically significant difference between the two benchmark measurements.
    The values are different enough that the most probable explanation is that they're random samples from two different data sets.
    t-statistic : +3.174          degrees of freedom : 28       95% confidence critical value : +2.042

 2: There's a statistically significant difference between the two benchmark measurements.
    The values are different enough that the most probable explanation is that they're random samples from two different data sets.
    t-statistic : +4.176          degrees of freedom : 28       95% confidence critical value : +2.042

 3: There's a statistically significant difference between the two benchmark measurements.
    The values are different enough that the most probable explanation is that they're random samples from two different data sets.
    t-statistic : +3.606          degrees of freedom : 28       95% confidence critical value : +2.042

 4: There's a statistically significant difference between the two benchmark measurements.
    The values are different enough that the most probable explanation is that they're random samples from two different data sets.
    t-statistic : +2.566          degrees of freedom : 28       95% confidence critical value : +2.042

 5: There's a statistically significant difference between the two benchmark measurements.
    The values are different enough that the most probable explanation is that they're random samples from two different data sets.
    t-statistic : +2.736          degrees of freedom : 28       95% confidence critical value : +2.042

 6: A human should check that the measurements for this metric look like random samples around a certain value to ensure the validity of this result.
```

Were you running benchmarks with the `DOCC_HTML_DIR` variable set?

Reply:
Yes I was. That metrics above is looking suspiciously uniform with a 5% improvement across the board. Since this is only affecting rendering I would expect bundle-registration and finalize-navigation-index to be unaffected. Since they also got 5% faster makes me think that the machine was under different load when collecting the different benchmark results.

Reply:
Hm. I'll try again...

Reply:
Okay re-running without using the computer during benchmarking and got no change:

```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Metric                                   │ Change          │ 82b1d4aeb2fb4afd8d5bc09893ac07b517999d89 │ ed63fcff152f11278632434b7f5d0636803b4403 │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Duration for 'bundle-registration'       │ no change¹      │ 3.803 sec                                │ 3.787 sec                                │
│ Duration for 'convert-total-time'        │ no change²      │ 6.542 sec                                │ 6.551 sec                                │
│ Duration for 'documentation-processing'  │ no change³      │ 2.513 sec                                │ 2.545 sec                                │
│ Duration for 'finalize-navigation-index' │ no change⁴      │ 0.104 sec                                │ 0.098 sec                                │
│ Peak memory footprint                    │ no change⁵,⁶    │ 425.8 MB                                 │ 425 MB                                   │
│ Data subdirectory size                   │ no change       │ 111.9 MB                                 │ 111.9 MB                                 │
│ Index subdirectory size                  │ no change       │ 994 KB                                   │ 994 KB                                   │
│ Total DocC archive size                  │ no change       │ 149.4 MB                                 │ 149.4 MB                                 │
│ Topic Anchor Checksum                    │ no change       │ c8f03d4e81da8e3b6d7fb7f63c771147         │ c8f03d4e81da8e3b6d7fb7f63c771147         │
│ Topic Graph Checksum                     │ no change       │ edce22bce4f8de475a5e58a2e45dcdc0         │ edce22bce4f8de475a5e58a2e45dcdc0         │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

 1: No statistically significant difference.
    The values are similar enough that the most probable explanation is that they're random samples from the same data set.
    t-statistic : +1.506          degrees of freedom : 28       95% confidence critical value : +2.042

 2: No statistically significant difference.
    The values are similar enough that the most probable explanation is that they're random samples from the same data set.
    t-statistic : -0.132          degrees of freedom : 28       95% confidence critical value : +2.042

 3: No statistically significant difference.
    The values are similar enough that the most probable explanation is that they're random samples from the same data set.
    t-statistic : -0.524          degrees of freedom : 28       95% confidence critical value : +2.042

 4: No statistically significant difference.
    The values are similar enough that the most probable explanation is that they're random samples from the same data set.
    t-statistic : +1.384          degrees of freedom : 28       95% confidence critical value : +2.042

 5: No statistically significant difference.
    The values are similar enough that the most probable explanation is that they're random samples from the same data set.
    t-statistic : +0.921          degrees of freedom : 28       95% confidence critical value : +2.042

 6: A human should check that the measurements for this metric look like random samples around a certain value to ensure the validity of this result.
```

Reply:
Since this "fix" was performing inconsistently I removed it and replaced it with a change to not write RenderJSON atomically. I added rationale for that change in the commit message: https://github.com/apple/swift-docc/pull/166/commits/4c30e3e87b4a90ff5e1e734961a71bc3510c3b98